app/src/main/java/com/example/android/codelabs/paging/ui/SearchRepositoriesViewModel.kt

```diff
@@ -36,11 +36,23 @@ class SearchRepositoriesViewModel(private val repository: GithubRepository) : Vi
         private const val VISIBLE_THRESHOLD = 5
     }
 
+    private val _repoLoadStatus = MutableLiveData<LoadState>()
+    val repoLoadStatus: LiveData<LoadState>
+        get() = _repoLoadStatus.distinctUntilChanged()
+
     private val queryLiveData = MutableLiveData<String>()
     val repoResult: LiveData<RepoSearchResult> = queryLiveData.switchMap {
+        _repoLoadStatus.value = LoadState.Loading
         liveData {
             val repos = repository.getSearchResultStream(it).asLiveData(Dispatchers.Main)
             emitSource(repos)
+        }.map {
+            // update the load status based on the result type
+            when (it) {
+                is RepoSearchResult.Success -> _repoLoadStatus.value = LoadState.Done
+                is RepoSearchResult.Error -> _repoLoadStatus.value = LoadState.Error(it.error)
+            }
+            it
```

Comment:
map with a side effect feels weird, but looks like there's no better operator :P

Reply:
I agree! It felt strange for me writing it as well :(

Reply:
@florina-muntenescu @ChrisCraik I think this would be achievable by adding a new `Observer` and `observeForever` on `repoResult`. The observer would then be removed and nullified in `onCleared`. Just tried it on this branch and it seems to be behaving the same.


```kotlin
    val repoResult: LiveData<RepoSearchResult> = queryLiveData.switchMap { queryString ->
        liveData {
            val repos = repository.getSearchResultStream(queryString).asLiveData(Dispatchers.Main)
            emitSource(repos)
        }
    }
    private var repoResultObserver: Observer<RepoSearchResult>? = null

    init {
        repoResultObserver = Observer<RepoSearchResult> {
            _repoLoadStatus.value = when (it) {
                is RepoSearchResult.Success -> LoadState.Done
                is RepoSearchResult.Error -> LoadState.Error(it.error)
            }
        }.also {
            repoResult.observeForever(it)
        }
    }

    override fun onCleared() {
        super.onCleared()
        repoResultObserver?.let { repoResult.removeObserver(it) }
        repoResultObserver = null
    }

```